docs/content/sips/xxx-manifest-redesign.md

```diff
@@ -0,0 +1,123 @@
+date = "2022-05-20T13:22:30Z"
+---
+
+Summary: A new design for Application Manifests (`spin.toml`)
+
+Owner: lann.martin@fermyon.com
+
+Created: May 20, 2022
+
+## Background
+
+Spin's Application Manifest file (usually named `spin.toml`) describes a Spin application's metadata,
+components, and trigger configurations. The current design supports a single trigger type and combines
+component and trigger configuration with an implicit one-to-one relationship:
+
+```toml
+name = "hello-world"
+trigger = { type = "http", base = "/hello" }
+
+[[component]]
+id = "world"
+# ...
+[component.trigger]
+route = "/world"
+```
+
+In order to support future features, we would like to enable more flexible trigger configurations:
+
+- Multiple trigger types in an application
+- Multiple triggers associated with a single component
+- One trigger associated with multiple components
+
+## Proposal
+
+### Global trigger config
+
+In order to allow for multiple trigger types, we remove the top-level `trigger.type` field.
+We also move all application metadata from the top-level config to a new `[application]`
+section, and all "global trigger config" to new `[application.trigger.<type>]` sections:
+
+```toml
+[application]
+name = "hello-world"
+[application.trigger.http]
+base = "/hello"
+```
+
+> In the short term we can continue to enforce the use of exactly one trigger type in an
+> application as a manifest validation step.
+
+### Trigger config
+
+In order to decouple triggers and components, we move triggers to new top-level
+`[[trigger.<type>]]` sections:
+
+```toml
+[[trigger.http]]
+route = "/world"
+component = "world"
+
+[[component]]
+id = "world"
+# ...
+```
+
+### Manifest version
+
+Since this proposal represents a backward-incompatible change to the manifest format, the manifest
+version changes. The existing `spin_version = "1"` field name leaves some room for misinterpretation
+as "compatible with Spin project version 1". As an optional part of this proposal, change that name:
+
+```toml
+spin_manifest_version = "2"
+```
+
+## Backward compatibility
+
+Existing "version 1" manifests can be transformed into this "version 2" data structure losslessly:
+
+```toml
+trigger = { type = "http", base = "/" }
+[[component]]
+id = "hello"
+[component.trigger]
+route = "/hello"
+```
+
+is equivalent to:
+
+```toml
+[application.trigger.http]
+base = "/"
+[[trigger.http]]
+route = "/hello"
+component = "hello"
+[[component]]
+id = "hello"
+```
+
+## Design options
+
+### Don't update the manifest version
+
+Instead of updating the manifest version we could not do that. It would be nice to include
+error message guidance and/or a `spin doctor` tool to upgrade the manifest format. This
+would be easy to detect if we switch the `spin_version` key.
+
+### Trigger IDs
+
+We might want to be able to reference individual triggers, for example to implement a
+`spin up --trigger=world`. It would save some compatibility headache if we require IDs now:
```

Comment:
This feels a bit intrusive in the simple case (which is likely to be common especially amongst the vital tyre-kicker demographic).  Could we, say, plan to default the ID to the trigger type, and require IDs only if there are multiple triggers of the same type?

Reply:
Oh wait is this meant to refer to individual components under a trigger?  (or should I say trigger-config-to-component maps)  This is what the example syntax seems to show, but if so I fear we may have two meanings for the term 'trigger'.

Reply:
I think my "trigger" is equivalent to your "trigger config". Each trigger has a "type", and a "trigger executor" can serve multiple triggers of a given type.

Reply:
So, in the HTTP case, each route is considered a separate trigger?  (And for Redis it would be each channel, etc.?)

Reply:
Yep, and a single component may handle multiple triggers.

Reply:
This seems really useful especially for testing purposes but it would be really nice if we could make this optional so we can make "the mean time to dopamine" to understand a spin.toml file and spin up as low as possible.

Reply:
Yeah that's why this idea is down here and not up there :point_up_2:.  :slightly_smiling_face: